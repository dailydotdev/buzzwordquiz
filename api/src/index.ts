import fastify, { FastifyInstance } from 'fastify';
import fastJson from 'fast-json-stringify';
import fastifyHelmet from 'fastify-helmet';
import routes from './routes';
import auth from './middlewares/auth';

const stringifyHealthCheck = fastJson({
  type: 'object',
  properties: {
    status: {
      type: 'string',
    },
  },
});

export default function app(): FastifyInstance {
  const isProd = process.env.NODE_ENV === 'production';

  const app = fastify({
    logger: true,
    disableRequestLogging: true,
    trustProxy: isProd,
  });

  app.register(fastifyHelmet);
  app.register(auth);

  app.get('/health', (req, res) => {
    res.type('application/health+json');
    res.send(stringifyHealthCheck({ status: 'ok' }));
  });

  app.register(routes, { prefix: '/' });

  return app;
}
