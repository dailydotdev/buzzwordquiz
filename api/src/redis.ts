import Redis, { RedisOptions, Pipeline } from 'ioredis';

declare module 'ioredis' {
  interface Pipeline {
    sortedTrim: (key: string, value: string | number) => Pipeline;
    leaderboard: (key: string, value: string | number) => Pipeline;
  }
}

export const redisOptions: RedisOptions = {
  host: process.env.REDIS_HOST,
  port: parseInt(process.env.REDIS_PORT, 10),
  password: process.env.REDIS_PASS,
};

export const redis = new Redis(redisOptions);

redis.defineCommand('sortedTrim', {
  numberOfKeys: 1,
  lua: `
    local index = -(ARGV[1]+1)
    local keys = redis.call("zrange", KEYS[1], 0, index)
    redis.call("del", unpack(keys))
    redis.call("zrem", KEYS[1], unpack(keys))
  `,
});

redis.defineCommand('leaderboard', {
  numberOfKeys: 1,
  lua: `
    local keys = redis.call("zrevrange", KEYS[1], 0, ARGV[1])
    if next(keys) == nil then
      return {}
    end
    return redis.call("mget", unpack(keys))
  `,
});
