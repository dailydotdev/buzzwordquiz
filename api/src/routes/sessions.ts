import { FastifyInstance, FastifyRequest } from 'fastify';
import { v4 as uuidv4 } from 'uuid';
import jwt from 'jsonwebtoken';
import { promisify } from 'util';
import AnswerBodySchema from '../../schemas/answerBody.json';
import { AnswerBodySchema as AnswerBodySchemaInterface } from '../../types/answerBody';
import CompleteBodySchema from '../../schemas/completeBody.json';
import { CompleteBodySchema as CompleteBodySchemaInterface } from '../../types/completeBody';

import {
  generateLettersFromAnswer,
  getNextQuestion,
  getQuestionsCount,
  getWordsLengths,
  Question,
} from '../models/question';
import { redis } from '../redis';
import { CachedSession, sessionIdToKey } from '../models/session';
import { preValidationHookHandler } from 'fastify/types/hooks';

const jwtSign = promisify(jwt.sign);

const JWT_TTL = 10 * 60;
export const MAX_SKIPS = 3;
export const SESSION_DURATION = 90;
export const LEADERBOARD_SIZE = 100;

interface UnansweredQuestion {
  logo: string;
  letters: string[];
  words: number[];
}

interface Session {
  id: string;
  startedAt: Date;
  score: number;
  skips: number;
  maxSkips: number;
  duration: number;
  nextQuestion: UnansweredQuestion;
}

const getReplySessionFromCacheAndNextQuestion = (
  cachedSession: CachedSession,
  nextQuestion: Question,
): Session => ({
  id: cachedSession.id,
  startedAt: new Date(cachedSession.startedAt),
  score: cachedSession.score,
  skips: cachedSession.skips,
  maxSkips: cachedSession.maxSkips,
  duration: cachedSession.duration,
  nextQuestion: {
    logo: nextQuestion.logo,
    letters: generateLettersFromAnswer(nextQuestion.answer),
    words: getWordsLengths(nextQuestion.answer),
  },
});

const fetchNextQuestAndUpdateCache = async (
  request: FastifyRequest,
): Promise<Question> => {
  const sessionId = request.gameSession.id;
  const key = sessionIdToKey(sessionId);
  request.gameSession.previousQuestions.push(request.gameSession.nextAnswer);
  const nextQuestion = await getNextQuestion(
    request.gameSession.questionsCount,
    request.gameSession.previousQuestions,
  );
  request.gameSession.questionsCount -= 1;
  request.gameSession.nextAnswer = nextQuestion.answer;
  await redis.set(key, JSON.stringify(request.gameSession));
  return nextQuestion;
};

const validateSession: preValidationHookHandler = async (request, reply) => {
  if (
    !request.gameSession ||
    request.gameSession.ended ||
    Date.now() - new Date(request.gameSession.startedAt).getTime() >
      (request.gameSession.duration + 1) * 1000
  ) {
    return reply.status(401).send();
  }
};

export default async function (fastify: FastifyInstance): Promise<void> {
  fastify.post(
    '/',
    async (): Promise<{ token: string; session: Session }> => {
      const sessionId = uuidv4();
      const startedAt = new Date();
      const token = (await jwtSign(
        { sessionId, exp: Math.floor(Date.now() / 1000) + JWT_TTL },
        process.env.JWT_KEY,
      )) as string;
      const questionsCount = await getQuestionsCount();
      const nextQuestion = await getNextQuestion(questionsCount, []);
      const key = sessionIdToKey(sessionId);
      const cachedSession: CachedSession = {
        id: sessionId,
        startedAt: startedAt.toISOString(),
        score: 0,
        skips: 0,
        maxSkips: MAX_SKIPS,
        duration: SESSION_DURATION,
        nextAnswer: nextQuestion.answer,
        previousQuestions: [],
        questionsCount: questionsCount - 1,
        ended: false,
      };
      await redis
        .pipeline()
        .set(key, JSON.stringify(cachedSession))
        .expire(key, JWT_TTL)
        .exec();
      return {
        token,
        session: getReplySessionFromCacheAndNextQuestion(
          cachedSession,
          nextQuestion,
        ),
      };
    },
  );

  fastify.post<{ Body: AnswerBodySchemaInterface }>(
    '/answer',
    {
      schema: {
        body: AnswerBodySchema,
      },
      preValidation: validateSession,
    },
    async (request): Promise<{ session?: Session; correct: boolean }> => {
      if (
        request.body.answer.toLowerCase() ===
        request.gameSession.nextAnswer.toLowerCase()
      ) {
        request.gameSession.score += 1;
        const nextQuestion = await fetchNextQuestAndUpdateCache(request);
        return {
          correct: true,
          session: getReplySessionFromCacheAndNextQuestion(
            request.gameSession,
            nextQuestion,
          ),
        };
      }
      return { correct: false };
    },
  );

  fastify.post(
    '/skip',
    {
      preValidation: validateSession,
    },
    async (request, reply): Promise<{ session: Session }> => {
      if (request.gameSession.skips >= request.gameSession.maxSkips) {
        return reply.status(400).send();
      }
      request.gameSession.skips += 1;
      const nextQuestion = await fetchNextQuestAndUpdateCache(request);
      return {
        session: getReplySessionFromCacheAndNextQuestion(
          request.gameSession,
          nextQuestion,
        ),
      };
    },
  );

  fastify.post<{ Body: CompleteBodySchemaInterface }>(
    '/complete',
    {
      schema: {
        body: CompleteBodySchema,
      },
      preValidation: async (request, reply) => {
        if (!request.gameSession || request.gameSession.ended) {
          return reply.status(401).send();
        }
      },
    },
    async (request): Promise<{ sessionId: string; score: number }> => {
      const sessionId = request.gameSession.id;
      const key = sessionIdToKey(sessionId);
      request.gameSession.ended = true;
      const leaderboardKey = `leaderboard:${sessionId}`;
      let pipeline = redis
        .pipeline()
        .set(key, JSON.stringify(request.gameSession));
      if (request.body.name) {
        pipeline = pipeline
          .set(
            leaderboardKey,
            JSON.stringify({
              id: sessionId,
              name: request.body.name,
              date: request.gameSession.startedAt,
              score: request.gameSession.score,
            }),
          )
          .zadd(
            'leaderboard.score.index',
            request.gameSession.score,
            leaderboardKey,
          )
          .sortedTrim('leaderboard.score.index', LEADERBOARD_SIZE);
      }
      await pipeline.exec();

      return {
        sessionId,
        score: request.gameSession.score,
      };
    },
  );
}
