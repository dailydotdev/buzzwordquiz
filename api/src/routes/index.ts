import { FastifyInstance } from 'fastify';

import sessions from './sessions';
import leaderboard from './leaderboard';

export default async function (fastify: FastifyInstance): Promise<void> {
  fastify.register(sessions, { prefix: '/sessions' });
  fastify.register(leaderboard, { prefix: '/leaderboard' });
}
