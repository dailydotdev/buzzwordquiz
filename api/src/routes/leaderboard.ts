import { FastifyInstance } from 'fastify';
import { redis } from '../redis';
import { LEADERBOARD_SIZE } from './sessions';

type LeaderboardRow = {
  id: string;
  name: string;
  date: string;
  score: number;
};

export default async function (fastify: FastifyInstance): Promise<void> {
  fastify.get(
    '/',
    async (): Promise<{ leaderboard: LeaderboardRow[] }> => {
      const rawResponse = await redis
        .pipeline()
        .leaderboard('leaderboard.score.index', LEADERBOARD_SIZE)
        .exec();
      return {
        leaderboard: rawResponse[0][1].map(
          (res) => JSON.parse(res) as LeaderboardRow,
        ),
      };
    },
  );
}
