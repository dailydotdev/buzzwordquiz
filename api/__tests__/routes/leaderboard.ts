import request from 'supertest';
import { FastifyInstance } from 'fastify';
import appFunc from '../../src';
import { LEADERBOARD_SIZE } from '../../src/routes/sessions';
import { redis } from '../../src/redis';

let app: FastifyInstance;

beforeAll(async () => {
  app = await appFunc();
  return app.ready();
});

afterAll(() => app.close());

const date = new Date(0);

describe('GET /leaderboard', () => {
  const path = '/leaderboard';

  it('should truncate leaderboard and return sorted by score', async () => {
    const pipeline = [...new Array(LEADERBOARD_SIZE + 20)].reduce(
      (pipeline, value, index) => {
        const key = `leaderboard:${index}`;
        return pipeline
          .set(
            key,
            JSON.stringify({
              id: index,
              name: `Leader${index}`,
              date: date.toISOString(),
              score: index + 1,
            }),
          )
          .zadd('leaderboard.score.index', index + 1, key);
      },
      redis.pipeline(),
    );
    await pipeline.exec();

    const res = await request(app.server).get(path).expect(200);
    expect(res.body).toMatchSnapshot();
  });

  it('should return leaderboard even if not fully filled', async () => {
    const pipeline = [...new Array(5)].reduce((pipeline, value, index) => {
      const key = `leaderboard:${index}`;
      return pipeline
        .set(
          key,
          JSON.stringify({
            id: index,
            name: `Leader${index}`,
            date: date.toISOString(),
            score: index + 1,
          }),
        )
        .zadd('leaderboard.score.index', index + 1, key);
    }, redis.pipeline());
    await pipeline.exec();

    const res = await request(app.server).get(path).expect(200);
    expect(res.body).toMatchSnapshot();
  });

  it('should return empty leaderboard', async () => {
    const res = await request(app.server).get(path).expect(200);
    expect(res.body).toMatchSnapshot({ leaderboard: [] });
  });
});
