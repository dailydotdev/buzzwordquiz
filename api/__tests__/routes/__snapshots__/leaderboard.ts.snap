// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /leaderboard should return empty leaderboard 1`] = `
Object {
  "leaderboard": Array [],
}
`;

exports[`GET /leaderboard should return leaderboard even if not fully filled 1`] = `
Object {
  "leaderboard": Array [
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 4,
      "name": "Leader4",
      "score": 5,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 3,
      "name": "Leader3",
      "score": 4,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 2,
      "name": "Leader2",
      "score": 3,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 1,
      "name": "Leader1",
      "score": 2,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 0,
      "name": "Leader0",
      "score": 1,
    },
  ],
}
`;

exports[`GET /leaderboard should truncate leaderboard and return sorted by score 1`] = `
Object {
  "leaderboard": Array [
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 119,
      "name": "Leader119",
      "score": 120,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 118,
      "name": "Leader118",
      "score": 119,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 117,
      "name": "Leader117",
      "score": 118,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 116,
      "name": "Leader116",
      "score": 117,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 115,
      "name": "Leader115",
      "score": 116,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 114,
      "name": "Leader114",
      "score": 115,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 113,
      "name": "Leader113",
      "score": 114,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 112,
      "name": "Leader112",
      "score": 113,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 111,
      "name": "Leader111",
      "score": 112,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 110,
      "name": "Leader110",
      "score": 111,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 109,
      "name": "Leader109",
      "score": 110,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 108,
      "name": "Leader108",
      "score": 109,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 107,
      "name": "Leader107",
      "score": 108,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 106,
      "name": "Leader106",
      "score": 107,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 105,
      "name": "Leader105",
      "score": 106,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 104,
      "name": "Leader104",
      "score": 105,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 103,
      "name": "Leader103",
      "score": 104,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 102,
      "name": "Leader102",
      "score": 103,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 101,
      "name": "Leader101",
      "score": 102,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 100,
      "name": "Leader100",
      "score": 101,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 99,
      "name": "Leader99",
      "score": 100,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 98,
      "name": "Leader98",
      "score": 99,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 97,
      "name": "Leader97",
      "score": 98,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 96,
      "name": "Leader96",
      "score": 97,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 95,
      "name": "Leader95",
      "score": 96,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 94,
      "name": "Leader94",
      "score": 95,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 93,
      "name": "Leader93",
      "score": 94,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 92,
      "name": "Leader92",
      "score": 93,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 91,
      "name": "Leader91",
      "score": 92,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 90,
      "name": "Leader90",
      "score": 91,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 89,
      "name": "Leader89",
      "score": 90,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 88,
      "name": "Leader88",
      "score": 89,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 87,
      "name": "Leader87",
      "score": 88,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 86,
      "name": "Leader86",
      "score": 87,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 85,
      "name": "Leader85",
      "score": 86,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 84,
      "name": "Leader84",
      "score": 85,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 83,
      "name": "Leader83",
      "score": 84,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 82,
      "name": "Leader82",
      "score": 83,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 81,
      "name": "Leader81",
      "score": 82,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 80,
      "name": "Leader80",
      "score": 81,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 79,
      "name": "Leader79",
      "score": 80,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 78,
      "name": "Leader78",
      "score": 79,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 77,
      "name": "Leader77",
      "score": 78,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 76,
      "name": "Leader76",
      "score": 77,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 75,
      "name": "Leader75",
      "score": 76,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 74,
      "name": "Leader74",
      "score": 75,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 73,
      "name": "Leader73",
      "score": 74,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 72,
      "name": "Leader72",
      "score": 73,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 71,
      "name": "Leader71",
      "score": 72,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 70,
      "name": "Leader70",
      "score": 71,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 69,
      "name": "Leader69",
      "score": 70,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 68,
      "name": "Leader68",
      "score": 69,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 67,
      "name": "Leader67",
      "score": 68,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 66,
      "name": "Leader66",
      "score": 67,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 65,
      "name": "Leader65",
      "score": 66,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 64,
      "name": "Leader64",
      "score": 65,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 63,
      "name": "Leader63",
      "score": 64,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 62,
      "name": "Leader62",
      "score": 63,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 61,
      "name": "Leader61",
      "score": 62,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 60,
      "name": "Leader60",
      "score": 61,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 59,
      "name": "Leader59",
      "score": 60,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 58,
      "name": "Leader58",
      "score": 59,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 57,
      "name": "Leader57",
      "score": 58,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 56,
      "name": "Leader56",
      "score": 57,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 55,
      "name": "Leader55",
      "score": 56,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 54,
      "name": "Leader54",
      "score": 55,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 53,
      "name": "Leader53",
      "score": 54,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 52,
      "name": "Leader52",
      "score": 53,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 51,
      "name": "Leader51",
      "score": 52,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 50,
      "name": "Leader50",
      "score": 51,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 49,
      "name": "Leader49",
      "score": 50,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 48,
      "name": "Leader48",
      "score": 49,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 47,
      "name": "Leader47",
      "score": 48,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 46,
      "name": "Leader46",
      "score": 47,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 45,
      "name": "Leader45",
      "score": 46,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 44,
      "name": "Leader44",
      "score": 45,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 43,
      "name": "Leader43",
      "score": 44,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 42,
      "name": "Leader42",
      "score": 43,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 41,
      "name": "Leader41",
      "score": 42,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 40,
      "name": "Leader40",
      "score": 41,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 39,
      "name": "Leader39",
      "score": 40,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 38,
      "name": "Leader38",
      "score": 39,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 37,
      "name": "Leader37",
      "score": 38,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 36,
      "name": "Leader36",
      "score": 37,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 35,
      "name": "Leader35",
      "score": 36,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 34,
      "name": "Leader34",
      "score": 35,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 33,
      "name": "Leader33",
      "score": 34,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 32,
      "name": "Leader32",
      "score": 33,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 31,
      "name": "Leader31",
      "score": 32,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 30,
      "name": "Leader30",
      "score": 31,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 29,
      "name": "Leader29",
      "score": 30,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 28,
      "name": "Leader28",
      "score": 29,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 27,
      "name": "Leader27",
      "score": 28,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 26,
      "name": "Leader26",
      "score": 27,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 25,
      "name": "Leader25",
      "score": 26,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 24,
      "name": "Leader24",
      "score": 25,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 23,
      "name": "Leader23",
      "score": 24,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 22,
      "name": "Leader22",
      "score": 23,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 21,
      "name": "Leader21",
      "score": 22,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 20,
      "name": "Leader20",
      "score": 21,
    },
    Object {
      "date": "1970-01-01T00:00:00.000Z",
      "id": 19,
      "name": "Leader19",
      "score": 20,
    },
  ],
}
`;
